bpo-32996: The bulk of What's New in Python 3.7

elprans opened the pull request:
This represents the bulk of the content for What's New in Python 3.7.

<!-- issue-number: bpo-32996 -->
https://bugs.python.org/issue32996
<!-- /issue-number -->

elprans committed:
bpo-32996: The bulk of What's New in Python 3.7

This represents the bulk of the content for What's New in Python 3.7.

1st1 commented on Doc/library/ensurepip.rst:
```diff
@@ -78,9 +78,6 @@ options:
 
-.. versionchanged:: 3.7.0
-   The exit status is non-zero if the command fails.
-
```
Right, this was backported to 3.6 and 2.7.

1st1 commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+
+Improved conistency of exceptions raised by :func:`issubclass` when called
+with a non-class and an abstract base class as the first and second
+arguments, respectively.  (Contributed by Josh Bronson in :issue:`33018`.)
```
I'd also add a ref to optimizations section's paragraph that covers `ABCMeta` being rewritten in C.

andresdelfino replied:
Note the "conistency"

1st1 commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+(Contributed by Yury Selivanov in :issue:`32251`.)
+
+The new :func:`asyncio.get_running_loop` function returns the currently
+running loop, or ``None`` if no loop is running.  This is in contrast with
```
This needs to be fixed: if no loop is running it raises a `RuntimeError`.

1st1 commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+:meth:`Future.get_loop() <asyncio.Future.get_loop>` methods
+return the instance of the loop on which a task or a future were created.
+(Contributed by Yury Selivanov in :issue:`32415`.)
+:meth:`Server.get_loop() <asyncio.Server.get_loop>` allows to do the same for
```
Need a new line here.

vadmium replied:
As far as I know _allows to_ is awkward and incorrect grammar in English, though it crops up often. In this case I suggest something like “_Server.get_loop_ allows doing the same . . .”, or even “_Server.get_loop_ does the same”.

1st1 commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+(Jelle Zijlstra with further improvements by INADA Naoki, Serhiy Storchaka,
+and Raymond Hettinger in :issue:`28638`.)
+
+Additionally, ``collections.namedtuple()`` now supports default values.
```
I'd remove "Additionally", feels a bit out of place.

1st1 commented on Doc/whatsnew/3.7.rst:
I'd also mention here that asyncio now uses this new mechanism, and that this new API is intended to replace `sys.set_coroutine_wrapper()` API, which is, btw, deprecated in 3.7.

1st1 commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+
+C API Changes
+=============
+
```
Need to mention PEP 567 related changes to the C API: https://www.python.org/dev/peps/pep-0567/#c-api

elprans replied:
These C APIs seem to be undocumented, BTW.

elprans replied:
I will add an entry together with the documentation for PEP 567 C API in a separate PR.

ilevkivskyi commented on Doc/whatsnew/3.7.rst:
```diff
@@ -768,17 +1490,68 @@ Add new clock identifiers:
+dynamically as specified by :pep:`560`.
+(Contributed by Ivan Levkivskyi in :issue:`32717`.)
+
+
```
There are some perf improvements in `typing` module due to PEP 560, I would probably mention 6-7x faster import of `typing`, but there are other perf improvements. They are all in the same PR https://github.com/python/cpython/pull/4906

serhiy-storchaka replied:
Perf improvements should be in the "Optimizations" section.

bbayles commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+ipaddress
+---------
+
+The new ``subnet_of`` and ``superset_of`` methods of
```
I think this ought to be `supernet_of` rather than `superset_of`, yes?

bbayles commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+
+:class:`~zipfile.ZipFile` now accepts the new *compresslevel* parameter to
+control the compression level.
+(Contributed by Stephen J. Turnbull. in :issue:`28102`.)
```
This references the incorrect issue - it was actually [21417](https://bugs.python.org/issue21417), and the attribution is also scrambled.

elprans replied:
Fixed. Thanks for catching this.

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+:meth:`datetime.isoformat() <datetime.datetime.isoformat>`.
 (Contributed by Paul Ganssle in :issue:`15873`.)
 
+The :class:`tzinfo <datetime.tzinfo>` objects now supprt sub-minute offsets.
```
Maybe it would be more correct to say “The _tzinfo_ class now supports . . .”, unless there are only specific instances with the support.

carljm replied:
s/supprt/support/

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+---
+
+A single empty field is now always quoted when written into a CSV
+file.  This allows to distinguish an empty row from a row consisting of
```
This allows distinguishing . . .

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+--
+
+The new :func:`gc.freeze`` function allows freezing all objects tracked
+by the garbage collector and exclude them from future collections.
```
. . . and excluding them . . .

vadmium commented on Doc/whatsnew/3.7.rst:
I think it makes more sense with “an” than without. Another way of writing it could be “. . . requires a _libssl_ implementation compatible with Open SSL’s 1.0.2 or 1.1 APIs.”

But after reading further down, it says one of these versions is required to build (and presumably use) the SSL module in general. Which seems to make this paragraph redundant.

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -768,17 +1490,68 @@ Add new clock identifiers:
+:func:`timeit.timeit` autorange now uses a single loop iteration if
+the benchmark takes less than 10 seconds, instead of 10 iterations.
+``python3 -m timeit -s 'import time' 'time.sleep(1)'`` now takes 4 seconds
+instead of 40 seconds.
```
This rule doesn’t make sense. Maybe it should be to use a single iteration if a benchmark is _slower_ than some value (0.2 s?). Does running sleep(11) also do a single iteration?

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -53,20 +55,123 @@ For full details, see the :ref:`changelog <changelog>`.
    so it's worth checking back even after reading earlier versions.
```
Is this PR not the time to remove this "prerelease users" note yet?

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -158,80 +290,30 @@ effort will be made to add such support.
-that would act like a normal :meth:`__getattr__` method, except that it will
-be defined on module *instances*.
+Python 3.7 allows defining :meth:`__getattr__` on modules and will call
+it whenever a module attribute is accessed.
```
Slightly misleading; like object `__getattr__`, module `__getattr__` is called only if the attribute is otherwise not found. So it's not called "whenever a module attribute is accessed." Suggested re-word: "...and will call it whenever a module attribute is otherwise not found."

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+method now supports now supports passing a :term:`path-like objects <path-like object>`
+as a socket path.  (Contributed by Yury Selivanov in :issue:`32066`.)
+
+The :func:`asyncio.windows_utils.socketpair`` function has been removed,
```
extra backtick after `windows_utils.socketpair`

serhiy-storchaka replied:
In the section "API and Feature Removals".

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
-:func:`~contextlib.asynccontextmanager` and
-:class:`~contextlib.AbstractAsyncContextManager` have been added. (Contributed
+The new :func:`~contextlib.nullcontext` is a simpler and faster alternative
+to :class:`~contextlib.ExitStack` for handling simple context managers.
```
I don't think that "handling simple context managers" is an accurate description of what `nullcontext` is good for. It won't help you handle or implement a simple context manager, it's useful only as a no-op context managers. Suggested re-word: "... a simpler and faster no-op context manager than ExitStack."

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
 cProfile
 --------
 
 :mod:`cProfile` command line now accepts ``-m module_name`` as an alternative
```
Nit: for consistency with other entries, this sentence should start with "The ".

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+A single empty field is now always quoted when written into a CSV
+file.  This allows to distinguish an empty row from a row consisting of
+a single empty field.
+(Contributed by Licht Takeuchi. in :issue:`32255`.)
```
Remove period after contributor name

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+decimal
+-------
+
+The :mod:`decimal` now uses :ref:`context variables <whatsnew37-pep567>` to
```
missing the word "module" between "decimal" and "now"

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
 dis
 ---
 
 The :func:`~dis.dis` function now is able to
```
s/now is/is now/

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+the new *depth* parameter.
+(Contributed by Serhiy Storchaka in :issue:`11822`.)
+
+Functions in the :mod:`dis` module now works with asynchronous generators and
```
s/works/work/

serhiy-storchaka replied:
Not in this section.

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+gc
+--
+
+The new :func:`gc.freeze`` function allows freezing all objects tracked
```
Extra backtick after `freeze`

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+The new ``subnet_of`` and ``superset_of`` methods of
+:class:`ipaddress.IPv6Network` and :class:`ipaddress.IPv4Network` can
+be used for network containment tests.
+(Contributed by Michel Albert and Cheryl Sabella. in :issue:`20825`.)
```
Remove period after "Sabella"

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+abc
+---
+
+Improved conistency of exceptions raised by :func:`issubclass` when called
```
s/conistency/consistency/

willingc replied:
sp./ consistency

serhiy-storchaka replied:
I think this is not worth an entry in this section. If you with, it can be added in the section "Other Language Changes" or like.

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -768,17 +1490,68 @@ Add new clock identifiers:
+now accepts negative *limit*, truncating the result to the
+``abs(limit)`` oldest frames.  To get the old behaviour, use
+the new *most_recent_first* argument to ``Traceback.format()``.
+(Contributed by Jesse Bakker. in :issue:`32121`.)
```
Remove period after "Bakker"

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -837,6 +1630,13 @@ The initialization of the default warnings filters has changed as follows:
 (Contributed by Nick Coghlan and Victor Stinner in :issue:`20361`,
 :issue:`32043`, and :issue:`32230`)
```
Need a period to end this sentence

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+avoid creating bound method instances.
+(Contributed by Yury Selivanov and INADA Naoki in :issue:`26110`.)
+
+The :mod:`asyncio` received a number of notable optimizations for commonly
```
missing the word "module"

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+:meth:`dict.copy` is now up to 5.5 times faster.
+(Contributed by Yury Selivanov in :issue:`31179`.)
+
+:func:`hasattr` and :func:`getattr`` are now about 4 times faster when
```
extra backtick after `getattr`

carljm commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+The :meth:`~selectors.BaseSelector.modify` methods of classes
+:class:`selectors.EpollSelector`, :class:`selectors.PollSelector`
+and :class:`selectors.DevpollSelector` may be around 10% faster under
+heavy loads.  (Contributed by Giampaolo Rodola' in :issue:`30014`)
```
use correct accent in contributor's name instead of trailing apostrophe?

elprans replied:
This is how Giampaolo is spelling his name.

carljm reviewed (commented):
Rendered the updated whatsnew in Sphinx and read through it in a browser. Looks good, modulo comments.

What a great Python release!

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -106,19 +211,46 @@ legacy C locale remains active when the core interpreter is initialized.
+an easy and consistent way to enter the Python debugger.
+
+Built-in ``breakpoint()`` calls :func:`sys.breakpointhook`.  By default, this
+latter imports :mod:`pdb` and then calls ``pdb.set_trace()``, but by binding
```
~~latter~~

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -158,80 +290,30 @@ effort will be made to add such support.
+Python 3.7 allows defining :meth:`__getattr__` on modules and will call
+it whenever a module attribute is accessed.
+
+A typical example of where this may be useful is attribute access deprecation
```
... attribute access, deprecation...

elprans replied:
This actually means "deprecation warnings for attribute access". I'll reword for clarity.

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -240,12 +322,12 @@ module:
-provide nanosecond resolution: they return a number of nanoseconds as a Python
-``int``.
+The new functions are similar in function to the existing functions
+without the ``_ns`` suffix, but instead of returning fractional seconds,
```
...suffix. They differ by returning nanoseconds as integers, instead of fractional seconds.

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
-the user intermix options and positional arguments on the command line,
-as is possible in many unix commands.  It supports most but not all
-argparse features.  (Contributed by paul.j3 in :issue:`14191`.)
+The new :meth:`ArgumentParser.parse_intermixed_args() \
```
As done on line 589, it probably is more readable to avoid using `\` and put the method link all on one line.

vadmium replied:
If it is just the backslash that worries you, perhaps you can remove it but leave the line break in

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+ast
+---
+
+:meth:`ast.literal_eval()` is now more strict.  Addition and subtraction of
```
~~more strict~~ stricter.

serhiy-storchaka replied:
This will better look in the section "Changes in the Python API". But I'm not sure it is worth mentioning in this document at all. This change makes the implementation closer to the documentation and will affect too little code that use literal_eval() in non-purposed way.

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+:meth:`loop.create_unix_server() <asyncio.AbstractEventLoop.create_unix_server>`,
+as well as :meth:`Server.start_serving() <asyncio.Server.start_serving>`, and
+:meth:`Server.serve_forever() <asyncio.Server.serve_forever>`
+can be used to decouple server instantiaion and serving.  The new
```
sp. instantiation

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+
+Support for directly ``await``-ing instances of :class:`asyncio.Lock` and
+other asyncio synchronization primitives has been deprecated.  An
+asynchronous context manager has to be used in order to acquire and release
```
~~has to~~ must

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+The new :func:`gc.freeze`` function allows freezing all objects tracked
+by the garbage collector and exclude them from future collections.
+This can be used before a POSIX ``fork()`` call to make the GC copy-on-write
+friendly or to speed up collection.  The new :func:`gc.unfreeze` functions
```
Looks like there is an extra period before "The new..."

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+method can be used to replace the logger stream after handler creation.
+(Contributed by Vinay Sajip in :issue:`30522`.)
+
+It is now possible to specify keyword arguments to handler constructor in
```
Perhaps this should say _handler constructors_?

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1012,90 +1916,193 @@ Other CPython Implementation Changes
+===========================================
+
+Function :c:func:`PySlice_GetIndicesEx` is deprecated and replaced with
+a macro if ``Py_LIMITED_API`` is not set or set to the value between
```
... is not set, is set to a value in the range between ``0x03050400`` and ``0x03060000`` (not inclusive), or is ``0x03060100`` or higher.

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1129,15 +2136,39 @@ API and Feature Removals
+
+* Direct instantiation of :class:`ssl.SSLSocket` and :class:`ssl.SSLObject`
+  objects is now prohibited. The constructors were never documented, tested,
+  or designed as public constructors. Users were suppose to use
```
sp. supposed

willingc commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1229,34 +2263,35 @@ Changes in the Python API
 
-* ``repr`` for :exc:`BaseException` has changed not to include trailing comma
-  in the output. Mind that most exceptions are affected by this change.
+* ``repr`` for :exc:`BaseException` has changed not to include the trailing
```
~~has changed not to include~~ has changed to exclude

willingc reviewed (commented):
@elprans Great work on the What's New. :sunny: I've left a few comments about typos and spelling.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+aifc
+----
+
+Improved exceptions raised for invalid number of channels and sample
```
This change is not worth be mentioned in this document at all.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+intermixing options and positional arguments.
+(Contributed by paul.j3 in :issue:`14191`.)
+
+:mod:`argparse` subparsers are now required by default.  This matches the
```
This will better look in the section "Changes in the Python API".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+asyncio
+-------
+
+The new :term:`provisional <provisional api>` :func:`asyncio.run` function can
```
This is a very long list. It would be better to write it in more compact form. See how long list of functions were introduced in previous releases.

1st1 replied:
Well, asyncio always have a lot of stuff in what's new. Every change here is important, please keep the section as is.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+returns an absolute scheduled callback timestamp.
+(Contributed by Andrew Svetlov in :issue:`32741`.)
+
+TCP sockets on Linux are now created with ``TCP_NODELAY`` flag set by default.
```
Not in this section.

1st1 replied:
I like it in this section, Serhiy. This is a super important change and must be highlighted.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+(Contributed by Marat Sharafutdinov. in :issue:`31943`.)
+
+The :meth:`loop.create_unix_connection() <asyncio.AbstractEventLoop.create_unix_connection>`
+method now supports now supports passing a :term:`path-like objects <path-like object>`
```
`now supports now supports`

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+``True`` if the callback was cancelled.
+(Contributed by Marat Sharafutdinov. in :issue:`31943`.)
+
+The :meth:`loop.create_unix_connection() <asyncio.AbstractEventLoop.create_unix_connection>`
```
Not in this section (if it is worth be mentioned at all).

1st1 replied:
Again, I disagree.  Serhiy, I appreciate your suggestions, but they are *suggestions*. A lot of people asked me about pathlib support in asyncio, so in my experience, this change is something that people (a) wanted to be fixed, (b) need to know about.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+:keyword:`async`/:keyword:`await` syntax.
+(Contributed by Andrew Svetlov in :issue:`32193`.)
+
+Support for directly ``await``-ing instances of :class:`asyncio.Lock` and
```
In the "Deprecated" section.

1st1 replied:
Please no. Again, I'm big -1 on having 10 different places describing asyncio changes instead of one section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+information.
+(Contributed by Andrew Svetlov in :issue:`32253`.)
+
+The :meth:`ReadTransport.resume_reading() <asyncio.ReadTransport.resume_reading>`
```
Not in this section (if it is worth be mentioned at all).

1st1 replied:
Yes, it's worth to be mentioned. And in this section :)

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+method can be used to determine the reading state of the transport.
+(Contributed by Yury Selivanov in :issue:`32356`.)
+
+Exceptions occurring in cancelled tasks are no longer logged.
```
Not in this section (if it is worth be mentioned at all).

1st1 replied:
ditto

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+
+.. _whatsnew37-asyncio-optimizations:
+
+Optimizations
```
This is a list of modules, this section is misplaced.

1st1 replied:
Again, asyncio is important enough to let people know about changes in it.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+atexit
+------
+
+The :func:`~atexit.atexit` callbacks are now correctly tracked and called for
```
This is a bug fix. Not in this section (if it is worth be mentioned at all).

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+subinterpreters.
+(Contributed by Marcel Plch in :issue:`31901`.)
+
+The traceback for :exc:`SystemExit` raised in an ``atexit`` callback
```
This is a bug fix, it shouldn't be in this document.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -446,35 +476,304 @@ Other Language Changes
+bdb
+---
+
+Pdb and other debuggers dependent on :mod:`bdb` will correctly step over
```
This is a bug fix.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
 the customisation of the CSS classes in the produced HTML calendar.
 (Contributed by Oz Tiram in :issue:`30095`.)
 
+The :func:`~calendar.weekday` function now accepts any value for its *year*
```
Not in this section. This is a minor enhancement.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+argument (previously only values from ``1`` to ``9999`` were accepted).
+(Contributed by Mark Gollahon in :issue:`28281`.)
+
+:meth:`Calendar.itermonthdates <calendar.Calendar.itermonthdates>`
```
Mention new methods first. The change in `itermonthdates()` is rather a bug fix.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+collections
+-----------
+
+The :func:`collections.namedtuple` factory has been reimplemented to rely
```
Not in this section. It is worth mentioning in two sections: optimizations and removals.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+Additionally, ``collections.namedtuple()`` now supports default values.
+(Contributed by Raymond Hettinger in :issue:`32320`.)
+
+:class:`collections.ChainMap` now preserves the order of the underlying
```
Not in this section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+can now take the multiprocessing context via the new *mp_context* argument.
+(Contributed by Thomas Moreau in :issue:`31540`.)
+
+``executor.submit()`` now raises a :exc:`RuntimeError` if called during
```
Not in this section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+configparser
+------------
+
+The :class:`~configparser.ConfigParser`` constructor now uses ``read_dict()``
```
Not in this section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+dictionary are now being implicitly converted to strings.
+(Contributed by James Tocknell. in :issue:`23835`.)
+
+The *filename* parameter of the
```
Not in this section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+csv
+---
+
+A single empty field is now always quoted when written into a CSV
```
Not in this section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+dbm
+---
+
+:func:`dbm.dumb` now supports reading read-only files and no longer writes the
```
Not in this section. This should be in two sections: a part in minor enhancements, and a part in deprecations.

``` :mod:`dbm.dumb` ``` instead of ``` :func:`dbm.dumb` ```.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+(Contributed by Serhiy Storchaka in :issue:`28847`.)
+
+
+:func:`dbm.open` now encodes filenames using the filesystem encoding rather
```
This is a bug fix.

serhiy-storchaka reviewed (commented):
A have reviewed a small part of the PR and have many comments. Many changes should be in other sections, more brief, or not be added at all.

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1667,245 @@ argument to generate a compressed archive.  A command line option
+
+The :func:`math.erf` and :func:`math.erfc` functions now use the (faster)
+C library implementation on most platforms.
+(Contributed by Serhiy Storchaka in :issue:`26121`.)
```
Duplicate entry, although I prefer your version :)

vadmium commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1229,34 +2263,35 @@ Changes in the Python API
-* Due to internal changes in :mod:`socket` you won't be able to
-  :func:`socket.fromshare` a socket :func:`~socket.socket.share`-ed in older
-  Python versions.
+* Due to internal changes in :mod:`socket` calling :func:`socket.fromshare`
```
Needs a comma: “Due to changes in _socket_, calling _fromshare_ . . .”

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+equivalent to CR.
+(Contributed by Bo Bayles in :issue:`32304`.)
+
+The unused ``install_misc`` command has been removed.
```
In the "Removed" section.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+and ``platforms`` fields are not specified as a list or a string.
+(Contributed by Berker Peksag in :issue:`19610`.)
+
+The ``upload`` command no longer tries to change CR end-of-line characters
```
This is a bug fix.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+
+:class:`distutils.core.setup` now warns if the ``classifiers``, ``keywords``
+and ``platforms`` fields are not specified as a list or a string.
+(Contributed by Berker Peksag in :issue:`19610`.)
```
and Neil Schemenauer

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+therefore included in source distributions.
+(Contributed by Ryan Gonzalez in :issue:`11913`.)
+
+:class:`distutils.core.setup` now warns if the ``classifiers``, ``keywords``
```
In "Changes in the Python API".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+enum members.
+(Contributed by Ethan Furman in :issue:`31801`.)
+
+In Python 3.8, attempting to check for non-Enum objects in :class:`Enum`
```
In "Deprecated"

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+fpectl
+------
+
+The ``fpectl`` module has been removed.  It was never enabled by
```
In "Removed".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+(Contributed by Julien Palard in :issue:`31639`.)
+
+
+idlelib and IDLE
```
Since 3.4 IDLE changes are documented separately and should not be here. In particularly this change was backported to 3.7.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+:ref:`whatsnew37_importlib_resources`.
+(Contributed by Barry Warsaw, Brett Cannon in :issue:`32248`.)
+
+:func:`importlib.reload` now raises :exc:`ModuleNotFoundError` if the module
```
The following two changes should be in other section. Actually they are closer to bug fixes which was decided to not backport.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+to make the conversion use monetary thousands separators and
+grouping strings.  (Contributed by Garvit in :issue:`10379`.)
+
+The :func:`locale.format` function is deprecated.
```
In "Deprecated".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+
+The :func:`locale.format` function is deprecated.
+
+The :func:`locale.getpreferredencoding` function now always returns ``'UTF-8'``
```
Isn't this already documented in the PEP 540 section?

elprans committed:
bpo-32996: More What's New in Python 3.7 edits

Remove redundant content, reclassify some entries, fix typos and address
other review feedback.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+macpath
+-------
+
+The :mod:`macpath` is now deprecated and will be removed in Python
```
In "Deprecated".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+mimetypes
+---------
+
+The MIME type of .bmp has been changed from ``'image/x-ms-bmp'`` to
```
"Changes in the Python API".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
+the new :func:`~os.preadv` and :func:`~os.pwritev` functions.
+(Contributed by Pablo Galindo in :issue:`31368`.)
+
+The mode argument of :func:`os.makedirs` no longer affects the file
```
"Changes in the Python API"

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -483,138 +782,458 @@ The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
-the new functions :func:`~os.preadv` and :func:`~os.pwritev`. (Contributed by
-Pablo Galindo in :issue:`31368`.)
 
 Exposed the system call *posix_spawn* through the new function
```
This feature was removed in 3.7.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -652,103 +1288,189 @@ can be set within the scope of a group.
 ``'^$'`` or ``(?=-)`` that matches an empty string.
 (Contributed by Serhiy Storchaka in :issue:`25054`.)
 
+Regular expressions compiled with the :const:`re.LOCALE` flag no longer
```
"Changes in the Python API"

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -652,103 +1288,189 @@ can be set within the scope of a group.
+when the compiled regular expression is used.
+(Contributed by Serhiy Storchaka in :issue:`30215`.)
+
+:exc:`FutureWarning` is now emitted if a regular expression contains
```
"Changes in the Python API".

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -768,17 +1490,68 @@ Add new clock identifiers:
+timeit
+------
+
+:func:`timeit.timeit` autorange now uses a single loop iteration if
```
Actually many other changes were made in this issue.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -768,17 +1490,68 @@ Add new clock identifiers:
+types
+-----
+
+The new :class:`~types.WrapperDescriptorType`,
```
Merge the following two entries into a single paragraph.

bbayles commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1501,272 @@ argument to generate a compressed archive.  A command line option
+
+:class:`~zipfile.ZipFile` now accepts the new *compresslevel* parameter to
+control the compression level.
+(Contributed by Bo Bayle in :issue:`21417`.)
```
Whoops, looks like my last name is misspelled here?

elprans replied:
Ah, sorry. Will fix.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -559,271 +584,169 @@ Improved Modules
 
 The :func:`~binascii.b2a_uu` function now accepts an optional *backtick*
-keyword argument.  When it's true, zeros are represented by ``'`'``
+keyword argument.  When it's ``True``, zeros are represented by ``'`'``
```
Please keep "true". Any boolean true value switches on this feature, not just literal `True`.

It is common to write "if is true" for input arguments, but "return True" for output value.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
````diff
@@ -1129,15 +2017,47 @@ API and Feature Removals
+* The unused :mod:`distutils ```install_misc`` command has been removed.
+  (Contributed by Eric N. Vander Weele in :issue:`29218`.)
+
+* The ``fpectl`` module has been removed.  It was never enabled by
````
It may be worth to add a separate section for this. Seems this is the first module removal since 3.0.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1501,272 @@ argument to generate a compressed archive.  A command line option
+C library implementation on most platforms.
+(Contributed by Serhiy Storchaka in :issue:`26121`.)
+
+Many functions and methods in the :mod:`abc` module have been reimplemented
```
This change are already documented above.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1501,272 @@ argument to generate a compressed archive.  A command line option
+(Contributed by Jelle Zijlstra with further improvements by INADA Naoki,
+Serhiy Storchaka, and Raymond Hettinger in :issue:`28638`.)
+
+:meth:`date.fromordinal` and :meth:`date.fromtimestamp` are now up to
```
I suggest: ``` Alternate :class:`datetime.date` constructors :meth:`~datetime.date.fromordinal` and :meth:`~datetime.date.fromtimestamp` ... ```

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -865,143 +1501,272 @@ argument to generate a compressed archive.  A command line option
-
-* Trace hooks may now opt in to receiving ``opcode`` events from the interpreter
-  by setting the new ``f_trace_opcodes`` attribute to :const:`True` on the frame
+* Trace hooks may now opt out of receiving the ``line`` and ``opcode`` events
```
"line" may now be opted out, but "opcode" may now be opted in.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1012,90 +1777,213 @@ Other CPython Implementation Changes
   :issue:`32303`.
 
+* The :func:`locals` dictionary now displays in the lexical order that
+  variables were defined.  Previously, the order was reversed.
```
Actually the order was arbitrary before 3.6 (as the order of all dict items). I would remove the note about the reversed order.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1012,90 +1777,213 @@ Other CPython Implementation Changes
+dbm
+---
+
+:mod:`dbm.dumb` now supports reading read-only files and no longer writes the
```
The first part of this text is not related to deprecation.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1012,90 +1777,213 @@ Other CPython Implementation Changes
-- Using non-integer value for selecting a plural form in :mod:`gettext` is
-  now deprecated.  It never correctly worked. (Contributed by Serhiy Storchaka
-  in :issue:`28692`.)
+aifc
```
Is it worth to add separate subsections for modules? Most contain only a single entry. This makes the text less compact.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1157,6 +2077,13 @@ Changes in Python behavior
   exceptions.
   (Contributed by Yury Selivanov in :issue:`32670`.)
 
+* :keyword:`async` and :keyword:`await` names are now reserved keywords.
```
Please move this at the first place, as it is the most visible breaking change. Would be nice to add recommendations for solving problems caused by this change.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1214,7 +2137,7 @@ Changes in the Python API
   (Contributed by Serhiy Storchaka in :issue:`29192`.)
 
 * ``Module``, ``FunctionDef``, ``AsyncFunctionDef``, and
```
I suggest to move this entry to the beginning of the sections, as it breaks several third-party projects.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1289,6 +2213,10 @@ Changes in the Python API
   of escaping all characters other than ASCII letters, numbers, and ``'_'``.
   (Contributed by Serhiy Storchaka in :issue:`29995`.)
 
+* :func:`os.listdir` and :func:`os.scandir` now emit file name as bytes when
```
This is a bug fix.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -1308,85 +2236,128 @@ Changes in the Python API
+  dictionary are now being implicitly converted to strings.
+  (Contributed by James Tocknell in :issue:`23835`.)
+
+* :mod:`csv` will now always quote empty fields.  This allows distinguishing
```
This is a bug fix.
